airbyte-integrations/connectors/source-paypal-transaction/source_paypal_transaction/source.py

```diff
@@ -0,0 +1,407 @@
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+import time
+from abc import ABC
+from datetime import datetime, timedelta
+from typing import Any, Callable, Iterable, List, Mapping, MutableMapping, Optional, Tuple, Union
+
+import requests
+from airbyte_cdk.sources import AbstractSource
+from airbyte_cdk.sources.streams import Stream
+from airbyte_cdk.sources.streams.http import HttpStream
+from airbyte_cdk.sources.streams.http.auth import Oauth2Authenticator
+from dateutil.parser import isoparse
+
+
+def get_endpoint(is_sandbox: bool = False) -> str:
+    if is_sandbox:
+        endpoint = "https://api-m.sandbox.paypal.com"
+    else:
+        endpoint = "https://api-m.paypal.com"
+    return endpoint
+
+
+class PaypalTransactionStream(HttpStream, ABC):
+
+    page_size = "500"  # API limit
+
+    # Date limits are needed to prevent API error: Data for the given start date is not available
+    # API limit: (now() - start_date_min) < start_date < (now() - start_date_max)
+    start_date_min: Mapping[str, int] = {"days": 3 * 364}  # API limit - 3 years
+    start_date_max: Mapping[str, int] = {"hours": 0}
+
+    stream_slice_period: Mapping[str, int] = {"days": 1}  # max period is 31 days (API limit)
+
+    requests_per_minute: int = 30  # API limit is 50 reqs/min from 1 IP to all endpoints, otherwise IP is banned for 5 mins
+
+    def __init__(
+        self,
+        start_date: Optional[datetime] = None,
+        end_date: Optional[datetime] = None,
+        is_sandbox: bool = False,
+        config: Optional[Mapping[str, Any]] = None,
+        **kwargs,
+    ):
+
+        # Initiate data from config
+        if config:
+            start_date = isoparse(config["start_date"])
+            end_date = isoparse(config.get("end_date")) if config.get("end_date") else None
+            is_sandbox = config["is_sandbox"]
+
+        self.start_date = start_date
+        self.end_date = end_date
+        self.is_sandbox = is_sandbox
+
+        self._validate_input_dates()
+
+        super().__init__(**kwargs)
+
+    @property
+    def end_date(self):
+        """Return initiated end_date or now()"""
+        now = datetime.now().replace(microsecond=0).astimezone()
+        if not self._end_date or self._end_date > now:
+            # If no end_date or end_date is in future then return now:
+            return now
+        else:
+            return self._end_date
+
+    @end_date.setter
+    def end_date(self, value: datetime):
+        self._end_date = value
+
+    def _validate_input_dates(self):
```

Comment:
Is it the correct place and way to validate input params? Should we put additional logging messages